File: packages/flutterfire_gen/test/helper/fcm_token.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'fcm_token.dart';

// **************************************************************************
// Generator: FlutterFireGen
// **************************************************************************

class ReadFcmToken {
  const ReadFcmToken({
    required this.tokenAndDevices,
    required this.createdAt,
    required this.fcmTokenId,
    required this.path,
    required this.fcmTokenReference,
  });

  final List<TokenAndDevice> tokenAndDevices;

  final DateTime? createdAt;

  final String fcmTokenId;

  final String path;

  final DocumentReference<ReadFcmToken> fcmTokenReference;

  factory ReadFcmToken.fromJson(Map<String, dynamic> json) {
    final extendedJson = <String, dynamic>{
      ...json,
    };
    return ReadFcmToken(
      tokenAndDevices: _tokenAndDevicesConverter
          .fromJson(extendedJson['tokenAndDevices'] as List<dynamic>?),
      createdAt: (extendedJson['createdAt'] as Timestamp?)?.toDate(),
      fcmTokenId: extendedJson['fcmTokenId'] as String,
      path: extendedJson['path'] as String,
      fcmTokenReference:
          extendedJson['fcmTokenReference'] as DocumentReference<ReadFcmToken>,
    );
  }

  factory ReadFcmToken.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadFcmToken.fromJson(<String, dynamic>{
      ...data,
      'fcmTokenId': ds.id,
      'path': ds.reference.path,
      'fcmTokenReference': ds.reference.parent.doc(ds.id).withConverter(
            fromFirestore: (ds, _) => ReadFcmToken.fromDocumentSnapshot(ds),
            toFirestore: (obj, _) => throw UnimplementedError(),
          ),
    });
  }

  ReadFcmToken copyWith({
    List<TokenAndDevice>? tokenAndDevices,
    DateTime? createdAt,
    String? fcmTokenId,
    String? path,
    DocumentReference<ReadFcmToken>? fcmTokenReference,
  }) {
    return ReadFcmToken(
      tokenAndDevices: tokenAndDevices ?? this.tokenAndDevices,
      createdAt: createdAt ?? this.createdAt,
      fcmTokenId: fcmTokenId ?? this.fcmTokenId,
      path: path ?? this.path,
      fcmTokenReference: fcmTokenReference ?? this.fcmTokenReference,
    );
  }
}

class CreateFcmToken {
  const CreateFcmToken({
    required this.tokenAndDevices,
  });

  final FirestoreData<List<TokenAndDevice>> tokenAndDevices;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'tokenAndDevices':
          _tokenAndDevicesConverter.toJson(tokenAndDevices.actualValue),
      'createdAt': FieldValue.serverTimestamp(),
    };
    final jsonPostProcessors = <({String key, dynamic value})>[];
    return {
      ...json,
      ...Map.fromEntries(jsonPostProcessors
          .map((record) => MapEntry(record.key, record.value))),
    };
  }
}

class UpdateFcmToken {
  const UpdateFcmToken({
    this.tokenAndDevices,
    this.createdAt,
  });

  final FirestoreData<List<TokenAndDevice>>? tokenAndDevices;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      if (tokenAndDevices != null)
        'tokenAndDevices':
            _tokenAndDevicesConverter.toJson(tokenAndDevices!.actualValue),
      if (createdAt != null) 'createdAt': createdAt,
    };
    final jsonPostProcessors = <({String key, dynamic value})>[];
    return {
      ...json,
      ...Map.fromEntries(jsonPostProcessors
          .map((record) => MapEntry(record.key, record.value))),
    };
  }
}

class DeleteFcmToken {}

/// Provides a reference to the fcmTokens collection for reading.
final readFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('fcmTokens')
    .withConverter<ReadFcmToken>(
      fromFirestore: (ds, _) => ReadFcmToken.fromDocumentSnapshot(ds),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a fcmToken document for reading.
DocumentReference<ReadFcmToken> readFcmTokenDocumentReference({
  required String fcmTokenId,
}) =>
    readFcmTokenCollectionReference.doc(fcmTokenId);

/// Provides a reference to the fcmTokens collection for creating.
final createFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('fcmTokens')
    .withConverter<CreateFcmToken>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a fcmToken document for creating.
DocumentReference<CreateFcmToken> createFcmTokenDocumentReference({
  required String fcmTokenId,
}) =>
    createFcmTokenCollectionReference.doc(fcmTokenId);

/// Provides a reference to the fcmTokens collection for updating.
final updateFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('fcmTokens')
    .withConverter<UpdateFcmToken>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a fcmToken document for updating.
DocumentReference<UpdateFcmToken> updateFcmTokenDocumentReference({
  required String fcmTokenId,
}) =>
    updateFcmTokenCollectionReference.doc(fcmTokenId);

/// Provides a reference to the fcmTokens collection for deleting.
final deleteFcmTokenCollectionReference = FirebaseFirestore.instance
    .collection('fcmTokens')
    .withConverter<DeleteFcmToken>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a fcmToken document for deleting.
DocumentReference<DeleteFcmToken> deleteFcmTokenDocumentReference({
  required String fcmTokenId,
}) =>
    deleteFcmTokenCollectionReference.doc(fcmTokenId);

/// Manages queries against the fcmTokens collection.
class FcmTokenQuery {
  /// Fetches [ReadFcmToken] documents.
  Future<List<ReadFcmToken>> fetchDocuments({
    GetOptions? options,
    Query<ReadFcmToken>? Function(Query<ReadFcmToken> query)? queryBuilder,
    int Function(ReadFcmToken lhs, ReadFcmToken rhs)? compare,
  }) async {
    Query<ReadFcmToken> query = readFcmTokenCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [FcmToken] documents.
  Stream<List<ReadFcmToken>> subscribeDocuments({
    Query<ReadFcmToken>? Function(Query<ReadFcmToken> query)? queryBuilder,
    int Function(ReadFcmToken lhs, ReadFcmToken rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadFcmToken> query = readFcmTokenCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadFcmToken] document.
  Future<ReadFcmToken?> fetchDocument({
    required String fcmTokenId,
    GetOptions? options,
  }) async {
    final ds = await readFcmTokenDocumentReference(
      fcmTokenId: fcmTokenId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [FcmToken] document.
  Stream<ReadFcmToken?> subscribeDocument({
    required String fcmTokenId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readFcmTokenDocumentReference(
      fcmTokenId: fcmTokenId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [FcmToken] document.
  Future<DocumentReference<CreateFcmToken>> add({
    required CreateFcmToken createFcmToken,
  }) =>
      createFcmTokenCollectionReference.add(createFcmToken);

  /// Sets a [FcmToken] document.
  Future<void> set({
    required String fcmTokenId,
    required CreateFcmToken createFcmToken,
    SetOptions? options,
  }) =>
      createFcmTokenDocumentReference(
        fcmTokenId: fcmTokenId,
      ).set(createFcmToken, options);

  /// Updates a specific [FcmToken] document.
  Future<void> update({
    required String fcmTokenId,
    required UpdateFcmToken updateFcmToken,
  }) =>
      updateFcmTokenDocumentReference(
        fcmTokenId: fcmTokenId,
      ).update(updateFcmToken.toJson());

  /// Deletes a specific [FcmToken] document.
  Future<void> delete({
    required String fcmTokenId,
  }) =>
      deleteFcmTokenDocumentReference(
        fcmTokenId: fcmTokenId,
      ).delete();
}

File: packages/flutterfire_gen_ts/test/helper/fcm_token.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentReference,
    FieldValue,
    GeoPoint,
    QueryDocumentSnapshot,
    QuerySnapshot,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadFcmToken {
  constructor ({
    fcmTokenId,
    path,
    tokenAndDevices,
createdAt,
  }: {
    fcmTokenId: string
    path: string
    tokenAndDevices: TokenAndDevice[]
createdAt?: Date
  }) {
    this.fcmTokenId = fcmTokenId
    this.path = path
    this.tokenAndDevices = tokenAndDevices
this.createdAt = createdAt
  }

  readonly fcmTokenId: string

  readonly path: string

  readonly tokenAndDevices: TokenAndDevice[]

readonly createdAt?: Date

  private static fromJson(json: Record<string, unknown>): ReadFcmToken {
    return new ReadFcmToken({
      fcmTokenId: json['fcmTokenId'] as string,
      path: json['path'] as string,
      tokenAndDevices: (json['tokenAndDevices'] as unknown[]).map((e) => e as TokenAndDevice),
createdAt: (json['createdAt'] as FirebaseFirestore.Timestamp | undefined)?.toDate(),
    })
  }


  static fromDocumentSnapshot(
    ds: FirebaseFirestore.DocumentSnapshot
  ): ReadFcmToken {
      const data = ds.data()!
      const cleanedData: Record<string, unknown> = {}
      for (const [key, value] of Object.entries(data)) {
          cleanedData[key] = value === null ? undefined : value
      }
      return ReadFcmToken.fromJson({
          ...cleanedData,
          fcmTokenId: ds.id,
          path: ds.ref.path
      })
  }

}

export class CreateFcmToken {
  constructor({
    tokenAndDevices,
  }: {
    tokenAndDevices: TokenAndDevice[] | FieldValue
  }) {
    this.tokenAndDevices = tokenAndDevices
  }


  readonly tokenAndDevices: TokenAndDevice[] | FieldValue

  toJson(): Record<string, unknown> {
  return {
    tokenAndDevices: this.tokenAndDevices,
'createdAt': FieldValue.serverTimestamp(),
  };
}

}

export class UpdateFcmToken {
  constructor({
    tokenAndDevices,
createdAt,
  }: {
    tokenAndDevices?: TokenAndDevice[] | FieldValue
createdAt?: Date
  }) {
    this.tokenAndDevices = tokenAndDevices
this.createdAt = createdAt
  }


  readonly tokenAndDevices?: TokenAndDevice[] | FieldValue

readonly createdAt?: Date

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
toJson(): Record<string, any> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const json: Record<string, any> = {}
  if (this.tokenAndDevices != undefined) {
  json['tokenAndDevices'] = this.tokenAndDevices
}
if (this.createdAt != undefined) {
  json['createdAt'] = this.createdAt
}

  return json
}

}

export class DeleteFcmToken {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the fcmTokens collection for reading.
 */
export const readFcmTokenCollectionReference = 
db
.collection('fcmTokens').withConverter<ReadFcmToken>({
  fromFirestore: (ds: FirebaseFirestore.DocumentSnapshot): ReadFcmToken => {
    return ReadFcmToken.fromDocumentSnapshot(ds)
  },
  toFirestore: () => {
      throw new Error(`toFirestore is not implemented for ReadFcmToken`)
  }
})


/**
 * Provides a reference to a fcmToken document for reading.
 * @param fcmTokenId - The ID of the fcmToken document to read.
 */
export const readFcmTokenDocumentReference = ({
  fcmTokenId
}: {
  fcmTokenId: string
}): FirebaseFirestore.DocumentReference<ReadFcmToken> =>
    readFcmTokenCollectionReference.doc(fcmTokenId);


/**
 * Provides a reference to the fcmTokens collection for creating.
 */
export const createFcmTokenCollectionReference = 
db
.collection('fcmTokens').withConverter<CreateFcmToken>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for CreateFcmToken`)
  },
  toFirestore: (obj: CreateFcmToken): FirebaseFirestore.DocumentData => {
      return obj.toJson()
  }
})


/**
 * Provides a reference to a fcmToken document for creating.
 * @param fcmTokenId - The ID of the fcmToken document to read.
 */
export const createFcmTokenDocumentReference = ({
  fcmTokenId
}: {
  fcmTokenId: string
}): FirebaseFirestore.DocumentReference<CreateFcmToken> =>
    createFcmTokenCollectionReference.doc(fcmTokenId);


/**
 * Provides a reference to the fcmTokens collection for updating.
 */
export const updateFcmTokenCollectionReference = 
db
.collection('fcmTokens').withConverter<UpdateFcmToken>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for UpdateFcmToken`)
  },
  toFirestore: (obj: UpdateFcmToken): FirebaseFirestore.DocumentData => {
      return obj.toJson()
  }
})


/**
 * Provides a reference to a fcmToken document for updating.
 * @param fcmTokenId - The ID of the fcmToken document to read.
 */
export const updateFcmTokenDocumentReference = ({
  fcmTokenId
}: {
  fcmTokenId: string
}): FirebaseFirestore.DocumentReference<UpdateFcmToken> =>
    updateFcmTokenCollectionReference.doc(fcmTokenId);


/**
 * Provides a reference to the fcmTokens collection for deleting.
 */
export const deleteFcmTokenCollectionReference = 
db
.collection('fcmTokens').withConverter<DeleteFcmToken>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for DeleteFcmToken`)
  },
  toFirestore: (): FirebaseFirestore.DocumentData => {
    throw new Error(`toFirestore is not implemented for DeleteFcmToken`)
  }
})


/**
 * Provides a reference to a fcmToken document for deleting.
 * @param fcmTokenId - The ID of the fcmToken document to read.
 */
export const deleteFcmTokenDocumentReference = ({
  fcmTokenId
}: {
  fcmTokenId: string
}): FirebaseFirestore.DocumentReference<DeleteFcmToken> =>
    deleteFcmTokenCollectionReference.doc(fcmTokenId);



/**
 * Manages queries against the fcmTokens collection.
 */
export class FcmTokenQuery {
  /**
   * Fetches fcmToken documents.
   * @param queryBuilder - Function to modify the query.
   * @param compare - Function to sort the results.
   */
  async fetchDocuments({
      
      queryBuilder,
      compare
  }: {
      
      queryBuilder?: (
          query: FirebaseFirestore.Query<ReadFcmToken>
      ) => FirebaseFirestore.Query<ReadFcmToken>
      compare?: (lhs: ReadFcmToken, rhs: ReadFcmToken) => number
  }): Promise<ReadFcmToken[]> {
      let query: FirebaseFirestore.Query<ReadFcmToken> =
          readFcmTokenCollectionReference
      if (queryBuilder != undefined) {
          query = queryBuilder(query)
      }
      const qs: QuerySnapshot<ReadFcmToken> = await query.get()
      let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadFcmToken>) =>
          qds.data()
      )
      if (compare != undefined) {
          result = result.sort(compare)
      }
      return result
  }

  /**
   * Fetches a specific fcmToken document.
   * @param fcmTokenId - The ID of the fcmToken document to fetch.
   */
  async fetchDocument({
    
    fcmTokenId
  }: {
    
    fcmTokenId: string
  }): Promise<ReadFcmToken | undefined> {
    const ds = await readFcmTokenDocumentReference({
      
      fcmTokenId
    }).get()
    return ds.data()
  }

  /**
   * Adds a fcmToken document.
   * @param createFcmToken - The fcmToken details to add.
   */
  async add({
    
    createFcmToken
  }: {
    
    createFcmToken: CreateFcmToken
  }): Promise<DocumentReference<CreateFcmToken>> {
    return createFcmTokenCollectionReference.add(createFcmToken)
  }

  /**
   * Sets a fcmToken document.
   * @param fcmTokenId - The ID of the fcmToken document to set.
   * @param createFcmToken - The fcmToken details to set.
   * @param options - Options for the set operation.
   */
  async set({
      
      fcmTokenId,
      createFcmToken,
      options
  }: {
      
      fcmTokenId: string
      createFcmToken: CreateFcmToken
      options?: FirebaseFirestore.SetOptions
  }): Promise<WriteResult> {
      if (options == undefined) {
          return createFcmTokenDocumentReference({
            
            fcmTokenId
          }).set(createFcmToken)
      } else {
          return createFcmTokenDocumentReference({ 
            
            fcmTokenId 
            }).set(createFcmToken, options)
      }
  }

  /**
   * Updates a specific fcmToken document.
   * @param fcmTokenId - The ID of the fcmToken document to update.
   * @param updateFcmToken - The details for updating the fcmToken.
   */
  async update({
    
    fcmTokenId,
    updateFcmToken
  }: {
    
    fcmTokenId: string
    updateFcmToken: UpdateFcmToken
  }): Promise<WriteResult> {
      return updateFcmTokenDocumentReference({ 
        
        fcmTokenId 
      }).update(updateFcmToken.toJson())
  }

  /**
   * Deletes a specific fcmToken document.
   * @param fcmTokenId - The ID of the fcmToken document to delete.
   */
  async delete({
    
    fcmTokenId
  }: {
    
    fcmTokenId: string
  }): Promise<WriteResult> {
      return deleteFcmTokenDocumentReference({ 
        
        fcmTokenId 
      }).delete()
  }
}
